src/gluonts/torch/model/tft/estimator.py

```diff
@@ -0,0 +1,390 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+from typing import Any, Dict, Iterable, List, Optional
+
+import numpy as np
+import torch
+from gluonts.core.component import validated
+from gluonts.dataset.common import Dataset
+from gluonts.dataset.field_names import FieldName
+from gluonts.itertools import Cyclic, IterableSlice, PseudoShuffled
+from gluonts.model.forecast_generator import QuantileForecastGenerator
+from gluonts.time_feature import TimeFeature, time_features_from_frequency_str
+from gluonts.torch.model.estimator import PyTorchLightningEstimator
+from gluonts.torch.model.predictor import PyTorchPredictor
+from gluonts.torch.util import IterableDataset
+from gluonts.transform import (
+    AddObservedValuesIndicator,
+    AddConstFeature,
+    AddTimeFeatures,
+    AsNumpyArray,
+    Chain,
+    ExpectedNumInstanceSampler,
+    RemoveFields,
+    SelectFields,
+    SetField,
+    TestSplitSampler,
+    Transformation,
+    ValidationSplitSampler,
+    VstackFeatures,
+)
+from gluonts.transform.sampler import InstanceSampler
+from gluonts.transform.split import TFTInstanceSplitter
+from torch.utils.data import DataLoader
+
+from .lightning_module import TemporalFusionTransformerLightningModule
+from .module import TemporalFusionTransformerModel
+
+PREDICTION_INPUT_NAMES = [
+    "past_target",
+    "past_observed_values",
+    "past_feat_dynamic_real",
+    "past_feat_dynamic_cat",
+    "feat_dynamic_real",
+    "feat_dynamic_cat",
+    "feat_static_real",
+    "feat_static_cat",
+]
+
+TRAINING_INPUT_NAMES = PREDICTION_INPUT_NAMES + [
+    "future_target",
+    "future_observed_values",
+]
+
+
+class TemporalFusionTransformerEstimator(PyTorchLightningEstimator):
+    """
+    Estimator class to train a Temporal Fusion Transformer model, as described in [LAL+21]_.
+
+    Parameters
+    ----------
+    freq
+        Frequency of the data to train on and predict.
+    prediction_length
+        Length of the prediction horizon.
+    context_length
+        Number of previous time series values provided as input to the encoder.
+        (default: None, in which case context_length = prediction_length).
+    quantiles
+        List of quantiles that the model will learn to predict.
+        Defaults to [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]
+    num_heads
+        Number of attention heads in self-attention layer in the decoder.
+    hidden_dim
+        Size of the LSTM & transformer hidden states.
+    variable_dim
+        Size of the feature embeddings.
+    static_dims
+        Sizes of the real-valued static features.
+    dynamic_dims
+        Sizes of the real-valued dynamic features that are known in the future.
```

Comment:
I’m a bit confused: why does one need to specify multiple sizes?

Reply:
I've added an explanation in the docstring for `TemporalFusionTransformerEstimator`. Does this make sense?

```
    TFT internally performs feature selection when making forecasts. For this
    reason, the dimensions of real-valued features can be grouped together if
    they correspond to the same variable (e.g., treat weather features as a
    one feature and holiday indicators as another feature).

    For example, if the dataset contains key "feat_static_real" with shape
    [batch_size, 3], we can, e.g.,
    - set ``static_dims = [3]`` to treat all five dimensions as a single feature
    - set ``static_dims = [1, 1, 1]`` to treat each dimension as a separate feature
    - set ``static_dims = [2, 1]`` to treat the first two dims as a single feature

    See ``gluonts.torch.model.tft.TemporalFusionTransformerModel.input_shapes``
    for more details on how the model configuration corresponds to the expected
    input shapes.
```

This way the TFT estimator / model is compatible with the DeepAR schema if we set
```python
d_feat_dynamic_real = [num_feat_dynamic_real],
```